### PR TITLE
Avoid forcing lambda parameter types in overloading resolution

### PR DESCRIPTION
**Pull request by odersky**
When taking the `typedArgs` of a `FunProto` we now always avoid failing
on untyped parameters of lambdas. Instead we give the parameter a ? type
and continue.

This allows to use lambdas with untyped parameters in more situations
than before.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -288,21 +288,28 @@ object ProtoTypes {
...
+          case Some(untpd.Function(args, _)) if !force =>
+            // If force = false, assume what we know about the parameter types rather than reporting an error.
+            // That way we don't cause a "missing parameter" error in `typerFn(arg)`
+            val paramTypes = args map {
+              case ValDef(_, tpt, _) if !tpt.isEmpty => typer.typedType(tpt).typeOpt
+              case _ => WildcardType
+            }
+            targ = arg.withType(defn.FunctionOf(paramTypes, WildcardType))
+          case Some(_) if !force =>
+            targ = arg.withType(WildcardType)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: Why not do it for partial functions too?

File: compiler/src/dotty/tools/dotc/config/Config.scala [hunk 2]
```diff
@@ -161,7 +161,7 @@ object Config {
   final val showCompletions = false
 
   /** If set, enables tracing */
-  final val tracingEnabled = false
+  final val tracingEnabled = true
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/config/Config.scala]: Revert this change?

File: docs/docs/reference/changed-features/overload-resolution.md [hunk 3]
```diff
@@ -3,29 +3,72 @@ layout: doc-page
...
+The handling of function values with missing parameter types has been improved. We can now
+pass such values in the first argument list of an overloaded application, provided
+that the remaining parameters suffice for picking a variant of the overloaded function.
+For example, the following code compiles in Dotty, while it results in an
+missing parameter type error in Scala2:
+```scala
+  def f(x: Int, f: Int => Int) = f(x)
+  def f(x: String, f: String => String) = f(x)
+  f("a", _.length)
+```
```
Comment (liufengyun) [inline docs/docs/reference/changed-features/overload-resolution.md]: > We can now pass such values in the first argument list of an overloaded application

It seems it works for all argument lists. The following code compile:

``` scala
def f(n: Int)(x: Int, f: Int => Int) = f(x)
def f(n: Int)(x: String, f: String => String) = f(x)

def g(m: Int)(n: Int)(x: Int, f: Int => Int) = f(x)
def g(m: Int)(n: Int)(x: String, f: String => String) = f(x)

f(4)("abc", _.reverse)
f(10)(2, _ + 2)

g(3)(4)("abc", _.reverse)
g(5)(10)(2, _ + 2)
```

Comment (odersky) [reply docs/docs/reference/changed-features/overload-resolution.md]: Yes, that's not contradicted. The "can now pass in the first parameter list" is the new aspect of it.

File: compiler/src/dotty/tools/dotc/config/Config.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/config/Config.scala]: oops yes.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: That's actually already the case since partial functions are also accepted by the `functionWithUnknownParamType` extractor. A added a sentence to the doc page that clarifies this.

**Comment by odersky** [comment]
test performance please
